app/channels/api/v1/race_channel.rb

```diff
@@ -0,0 +1,298 @@
+      message = {
+        message:      'A new message has been posted',
+        chat_message: Api::V4::ChatMessageBlueprint.render_as_hash(chat_message)
+      }
+      if onsite
+        message[:chat_html] = ApplicationController.render(
+          partial: 'chat_messages/show',
+          locals:  {chat_message: chat_message}
+        )
+      end
+
+      broadcast_to(
+        @race,
+        Api::V1::WebsocketMessageBlueprint.render_as_hash(
+          Api::V1::WebsocketMessage.new('new_message', message)
+        )
+      )
+    else
+      transmit_user('message_creation_error', message.errors.full_messages.to_sentence)
+    end
+  rescue StandardError => e
+    Rails.logger.error([e.message, *e.backtrace].join($RS))
+    Rollbar.error(e, 'Uncaught error for Api::V1::RaceChannel#send_message')
+    transmit_user('fatal_error', 'A fatal error occurred while processing your message')
+  end
+
+  def attach_file
+    return unless @permitted
+    return if check_user
+    return if check_oauth
+
+    update_race_instance
+
+    broadcast_race_update('new_attachment', 'The race owner has attached a new file')
+  end
+
+  private
+
+  def update_race_instance
+    # Instance variables do not update automatically, so we call this function before anything that needs
+    # to check the state of the race variable to make sure it isn't stale
+    @race.reload
+  end
+
+  def broadcast_race_update(type, msg)
+    update_race_instance
+    msg = {
+      message: msg,
+      race:    Api::V4::RaceBlueprint.render_as_hash(@race, view: @race.type)
+    }
+    if onsite
+      msg[:entrants_html] = ApplicationController.render(partial: 'races/entrants_table', locals: {race: @race})
+      msg[:stats_html] = ApplicationController.render(partial: 'races/stats', locals: {race: @race})
+
+      if @race.randomizer?
+        msg[:attachments_html] = ApplicationController.render(partial: 'races/files', locals: {race: @race})
+      end
+    end
+
+    ws_msg = Api::V1::WebsocketMessage.new(type, msg)
+    broadcast_to(@race, Api::V1::WebsocketMessageBlueprint.render_as_hash(ws_msg))
+  end
+
+  # Starts the race if every entrant is readied up, otherwise does nothing
+  def maybe_start_race
+    update_race_instance
+    return if @race.started? || !@race.entrants.all?(&:ready?) || @race.entrants.count < 2
+
+    @race.update(started_at: Time.now.utc + 20.seconds, status: :in_progress)
+    broadcast_race_update('race_start_scheduled', 'Race starting soon')
+    GlobalRaceUpdateJob.perform_later('race_started', 'A race has started', @race)
+  end
+
+  # Ends the race if all entrants have either finished or forfeited, otherwise does nothing
+  def maybe_end_race
+    update_race_instance
+    return if !@race.started? || !@race.entrants.all?(&:done?)
+
+    @race.update(status: :ended)
+    broadcast_race_update('race_ended', 'All entrants have finished')
+    GlobalRaceUpdateJob.perform_later('race_finished', 'A race has finished', @race)
+  end
+end
```

Comment:
I don't know of any "accepted" solution to this.  I haven't really seen anything that wasn't a toy/trivial example of using action cable.  I think we kind of have free range to try things out and experiment.  I just think it might be beneficial to avoid having a ton of channels that need to be subscribed to if that is going to increase the complexity for a 3rd party dev.  Maybe we could leverage models more or something along those lines?

Related, but why are large files considered always bad?  I feel like having everything related to something all in one place would be a good thing since you wouldn't have to try and figure out a chain of callbacks or something along those lines

Reply:
That's a really good question! There are some practical disadvantages, like readability.
```ruby
class ThingsController < ApplicationController
  def index
  end

  def create
  end

  def show
  end
end
```
This controller clearly tells you what it does -- list, create, and show Things. But if it grows to 400 lines, it doesn't. You have to scroll a lot, have a good memory, good tooling, etc.

And there's a theoretical disadvantage -- large files can be indicative of *other* bad things.

e.g. Run started small but eventually grew a lot of ways to render itself. If we'd started Splits.io today and knew we wanted 19 ways of rendering a Run, we'd build the rendering engine to be independent from the start. But we didn't know that, we slowly grew it from nothing, and it's hard to see how complex it is when you're in the weeds.

But what you *can* see is the file size. When it starts getting long, it's a good idea to dive in a bit and see if you've overlooked whether a new model/controller/etc. wants to be split out. For us, we had controller actions for `download`, `run_history_csv`, and `segment_history_csv` bloating their controllers and kind of all doing the same thing -- exporting a run.

The weirdest part about this type of situation is none of us even realized how similar they were in UX -- the download button for the different timer types was completely separate from the "export to CSV" ones.

By forcing ourselves to think about whether there was a new controller begging to be created, we came up with Run Export, and alongside that came one button to handle all types of exports.
```
app/
  controllers/
    runs_controller.rb
    runs/
      exports_controller.rb
```
It also gave us a nice birds-eye view in the directory structure, just like we have in simple controllers. The code got cleaner and the user experience did as well.

I do agree channels will be a different beast, as splitting up one channel into two causes a very different experience for consumers of the channel, where in controllers it does not.

I do notice some repeated code in similar methods, e.g. ready/unready are basically identical. Could we share code between these methods more, or even merge them into one `set_ready(ready)`-esque method? How about for other entrant state-change type methods?

Reply:
Thanks for the info.  I can definitely see now why having concise files would make things better in the long run.

My only concern with combining the methods (which may be baseless) was that I wanted each action to be explicit.  You knew what you were doing by what method was being called.  Originally in SRR I had leaving a race before it started and forfeiting be the same method, and that seemed like it would be confusing to consumers.  You had to track the state of the race to know what the method was going to do.  I probably took that a little too far though here, since by passing it a value to some of the methods you maintain that explicit nature without needing to redefine everything.

Reply:
Yeah I totally agree we shouldn't use one "command" to mean different things based on state, like a `toggle_ready` or the example you mentioned. The consumer should know exactly what effect its command will have if successful, even if it is out of sync with the server state.

The current setup is great as an API design, but I do think we need to find a way to share some of the duplicated code, however that is. Like you said we should feel free to invent something.